nni/retiarii/oneshot/pytorch/differentiable.py

```diff
@@ -0,0 +1,325 @@
+
+        # phase 1: architecture step
+        # The _resample hook is kept for some darts-based NAS methods like proxyless.
+        # See code of those methods for details.
+        self._resample()
+        arc_optim.zero_grad()
+        arc_step_loss = self.model.training_step(val_batch, 2 * batch_idx)
+        if isinstance(arc_step_loss, dict):
+            arc_step_loss = arc_step_loss['loss']
+        self.manual_backward(arc_step_loss)
+        self.finalize_grad()
+        arc_optim.step()
+
+        # phase 2: model step
+        self._resample()
+        self.call_user_optimizers('zero_grad')
+        loss_and_metrics = self.model.training_step(trn_batch, 2 * batch_idx + 1)
+        w_step_loss = loss_and_metrics['loss'] \
+            if isinstance(loss_and_metrics, dict) else loss_and_metrics
+        self.manual_backward(w_step_loss)
+        self.call_user_optimizers('step')
+
+        self.call_lr_schedulers(batch_idx)
+
+        return loss_and_metrics
+
+    def _resample(self):
+        # Note: This hook is kept for following darts-based NAS algs.
+        pass
+
+    def finalize_grad(self):
+        # Note: This hook is currently kept for Proxyless NAS.
+        pass
+
+    @property
+    def default_replace_dict(self):
+        return {
+            LayerChoice : DartsLayerChoice,
+            InputChoice : DartsInputChoice
+        }
+
+    def configure_architecture_optimizers(self):
+        # The alpha in DartsXXXChoices is the architecture parameter of DARTS. All alphas share one optimizer.
+        ctrl_params = {}
+        for _, m in self.nas_modules:
+            if m.name in ctrl_params:
+                assert m.alpha.size() == ctrl_params[m.name].size(), 'Size of parameters with the same label should be same.'
+                m.alpha = ctrl_params[m.name]
+            else:
+                ctrl_params[m.name] = m.alpha
+        ctrl_optim = torch.optim.Adam(list(ctrl_params.values()), 3.e-4, betas=(0.5, 0.999),
+                                           weight_decay=1.0E-3)
+        return ctrl_optim
+
+
+class _ArchGradientFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, x, binary_gates, run_func, backward_func):
+        ctx.run_func = run_func
+        ctx.backward_func = backward_func
+
+        detached_x = x.detach()
+        detached_x.requires_grad = x.requires_grad
+        with torch.enable_grad():
+            output = run_func(detached_x)
+        ctx.save_for_backward(detached_x, output)
+        return output.data
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        detached_x, output = ctx.saved_tensors
+
+        grad_x = torch.autograd.grad(output, detached_x, grad_output, only_inputs=True)
+        # compute gradients w.r.t. binary_gates
+        binary_grads = ctx.backward_func(detached_x.data, output.data, grad_output.data)
+
+        return grad_x[0], binary_grads, None, None
+
+
+class ProxylessLayerChoice(nn.Module):
```

Comment:
I think the location of xxxChoice module is confusing now. ProxylessNAS xxxChoice are in differentiable.py, but Darts xxxChoice are in darts.py. Why?